external/implementations/samlconf-openam-impl/src/main/java/org/codice/compliance/saml/plugin/openam/OpenAMDataParser.java

```diff
@@ -0,0 +1,125 @@
+
+import static org.codice.compliance.CommonKt.USER_LOGIN;
+
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.xml.sax.SAXException;
+
+class OpenAMDataParser {
+
+  //  A class only for retrieving values from the openAM data file.
+  private final ImmutableMap<String, String> formMapping;
+  private final String openAMRealm;
+  private final String baseURL;
+
+  private static final String REALM = "realmName";
+  private static final String BASE_URL = "baseUrl";
+  private static final String LOGIN_FORM = "loginForm";
+  private static final String INPUT = "input-";
+  private static final String OUTPUT = "output-";
+
+  private static final String USER_ARG = "USER.ARG";
+  private static final String PASSWORD_ARG = "PASS.ARG";
+
+  private static final String fileName = "OpenAMLoginData.xml";
+
+  OpenAMDataParser() {
+
+    DocumentBuilder doc;
+    try {
+      DocumentBuilderFactory docBF = DocumentBuilderFactory.newInstance();
+      doc = docBF.newDocumentBuilder();
+    } catch (ParserConfigurationException e) {
+      throw new IllegalStateException(e);
+    }
+
+    Document xml;
+    try {
+      InputStream a = getClass().getClassLoader().getResourceAsStream(fileName);
+      if (a == null) {
+        throw new IllegalStateException(fileName + " was not found");
+      }
+      xml = doc.parse(a);
+
+    } catch (IOException | SAXException e) {
+      throw new IllegalStateException(e);
+    }
+
+    //  Parsing the XML login data
+    String tempRealm = xml.getElementsByTagName(REALM).item(0).getTextContent();
+    openAMRealm = (tempRealm.equals("/")) ? "" : tempRealm;
+    baseURL = xml.getElementsByTagName(BASE_URL).item(0).getTextContent();
+
+    //  Since the authentication payload sent is customizable by the client of OpenAM
+    //  This parsing has to be dynamic. Loop through the login form tag and continue while
+    //  both the Input and Output tags are supplied.
+    Element loginInfo = (Element) xml.getElementsByTagName(LOGIN_FORM).item(0);
+
+    Map<String, String> tempMap = new HashMap<>();
+    int i = 1;
+    while (hasChildTag(loginInfo, INPUT + i) && hasChildTag(loginInfo, OUTPUT + i)) {
+
+      tempMap.put(
+          loginInfo.getElementsByTagName(INPUT + i).item(0).getTextContent(),
+          expandArg(loginInfo.getElementsByTagName(OUTPUT + i).item(0).getTextContent()));
+      i++;
+    }
+
+    formMapping = ImmutableMap.copyOf(tempMap);
+  }
+
+  private boolean hasChildTag(Element element, String tagName) {
+    return element.getElementsByTagName(tagName).getLength() != 0;
```

Comment:
✏️ We might want to protect against `getElementsByTagName(tagName)` returning `null`. According to the `Element` interface, they're vague on whether or not the return could be `null`.

Reply:
After looking into it, the javadoc states it `returns a list of matching <code>Element</code> nodes`. If there is a bad request it returns an empty list.